src/features/tooltip/Tooltip/Tooltip.tsx

```diff
@@ -1,38 +1,51 @@
 import ReactMarkdown from 'react-markdown';
 import clsx from 'clsx';
 import { Close16 } from '@konturio/default-icons';
-import { useEffect, useState } from 'react';
+import { useState } from 'react';
 import { LinkRenderer } from '~components/LinkRenderer/LinkRenderer';
 import { parseLinksAsTags } from '~utils/markdown/parser';
 import s from './Tooltip.module.css';
 import type { TooltipData } from '~core/shared_state/currentTooltip';
 
+type Position = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
+type Coords = { x: number; y: number };
+
+function findTooltipPosition(clickPosition?: Coords | null): Position | null {
+  const { height, width } = window.visualViewport;
+  if (!clickPosition) return null;
+  // click was on the bottom right side
+  if (clickPosition.y > height / 2 && clickPosition.x > width / 2)
+    return 'top-left';
+  // click was on the bottom left side
+  if (clickPosition.y > height / 2) return 'top-right';
+  // click was on the top right side
+  if (clickPosition.x > width / 2) return 'bottom-left';
+  // click was on the top right side
+  return 'bottom-right';
+}
+
 export function Tooltip({
   properties,
   closeTooltip,
 }: {
   properties: TooltipData | null;
   closeTooltip: () => void;
 }) {
-  const [position, setPosition] = useState<
-    'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' | null
-  >(null);
+  const [position, setPosition] = useState<Position | null>(null);
+  const [prevCoords, setPrevCoords] = useState<Coords | null | undefined>(null);
 
-  useEffect(() => {
-    if (properties) {
-      const { height, width } = window.visualViewport;
-      // Case - click was on the bottom half, put tooltip on top half
-      if (properties.position.y > height / 2) {
-        // click was on the rigth side
-        if (properties.position.x > width / 2) setPosition('top-left');
-        else setPosition('top-right');
-      }
-      // click was on the top right side
-      else if (properties.position.x > width / 2) {
-        setPosition('bottom-left');
-      } else setPosition('bottom-right');
-    } else setPosition(null);
-  }, [properties?.position]);
+  if (
+    prevCoords?.x !== properties?.position.x &&
```

Comment:
we have `properties: TooltipData | null; ` 
TooltipData.position is required field. So we are sure, that we have it always. 
at line 61 we process case  `if (!properties) return null;`
I suggest to do it at start, right after `useState` line 35, and get rid of all these optional chaining checks: properties?.position.